packages/compiler/core/semantic-walker.ts

```diff
@@ -279,6 +279,20 @@ export function mapChildModels(program: Program): ReadonlyMap<ModelType, readonl
   return map;
 }
 
+const ChildModelMap = Symbol();
+/**
+ * @returns the mapping of parent model to child model.
+ * Will cache the value so it is computed only once.
```

Comment:
I worry about this because Program is mutable. How does one invalidate the cache?

Reply:
oh I see, it's on the program

Reply:
actually, still a potential problem since the program map isn't reset if program is mutated.

Reply:
yeah, I think we need to discuss on how to cache things in the program. Right now we still have the issue that if you call `getRoutes` twice it will recompute all the routes and emit the same diagnostics twice.

However would that be an issue? In `onValidate` or `onEmit` you shouldn't be mutating the program(You can right now but you shouldn't no?) I think if we really have something where we want to mutate the program we should have another hook that will be able to mutate and then we can invalidate all cache after/during it?

Reply:
You're right that you shouldn't mutate in onValidate or onEmit, but I was thinking more if a decorator decided to call this as decorators can add types currently.

This probably does need more discussion. I suppose it's possible to expose different surface to decorators and onXxx but also complex. The other extreme is to just live with it and say don't do that, but I worry about the footgun potential and it becomes really hard to document what you can't and can't do where.

Reply:
Could do that to start:
- Create a new `cache` entry in the program especially meant to cache things.
- We could clear the cache before onValidate and in between each decorator.
- This now introduce an potential performance issue if you think you can use some heavy cached value but it is not cached. So maybe we can have a flag to emit warning when using a cached value in a decorator so you can track those issues.

Or we skip the cache clear between decorator for now but now instead of a perf issue we could just get confusing bugs.

Reply:
I haven't thought about this issue super deeply but I am concerned about adding any caching here, and caching as a solution to duplicate errors feels like solving the problem at the wrong layer. Would love to discuss in more detail in a meeting.

Reply:
Created an issue for this https://github.com/microsoft/cadl/issues/368. Do you want to have a dedicated meeting for this or should we just discuss this in a design meeting. Feels like there is some brainstorming to do which might not be the right fit.

Reply:
Resolved via the derivedModels